### PR TITLE
Add the ability to select which functions or processes you which to extract capabilities from

### PR DESCRIPTION
**Pull request by yelhamer**
This PR adds the ability to select which function/process capa should extract capabilities from. The proposed syntax is as follows:
```bash
$ capa malware.exe --functions 0x645fa0,0x543dd0,0x630ac0 # static analysis
$ capa malware.log --processes 3288,4321,3234 # dynamic analysis
```

I haven't added a testcase for dynamic analysis. I am planning to do so once the Drakvuf feature extractor gets merged (#2143) since that's a big motive for this PR.

Nuance: I couldn't find the right names for some of the internal variables, so please feel free to set them as you wish.

Thanks :)

### Checklist

- [ ] No CHANGELOG update needed

- [ ] No new tests needed

- [ ] No documentation update needed

### CODE REVIEW

**Comment by williballenthin** [comment]
First, I think this is a very reasonable feature to add, especially with the Drakvuf sandbox support! I'm happy that we should be able to remove similar logic in `show-features` ([here](https://github.com/mandiant/capa/blob/1975b6455c9e5ccfdf295210659730c1cfc50d25/scripts/show-features.py#L146-L158)) with these changes. It seems like there are multiple places that this would fit in.

There are two things to discuss:
  1. the command line argument syntax, and
  2. the design of the filtering.

For (1), I don't imagine any major problems, though we may want to consider if these arguments are commonly used enough to be shown all the time, or only in an "expert" mode, or suggested when the Drakvuf sandbox is encountered, etc. That being said, let's be careful not to get derailed by tiny details and hold up the merge. If there's debate, the arguments could always be considered "experimental" for a bit until we stabilize them.

For (2), the question is: how do we filter the functions/scopes/features, particularly in a way that enables further extension, if necessary? That is, say we want to also filter on threads or basic blocks, could we build this easily?

The primary place to consider is the function signature to `find_capabilities` (and related routines). I am concerned about adding too many optional arguments that become difficult to reason about. I'd prefer the signature of this core routine to be as succinct as possible.

An alternative design is to introduce wrappers around the feature extractors that can filter the scopes/features. They would act just like the underlying feature extractor, but would yield only the entries that are requested. Then, the wrapped feature extractor can be passed around as-is today. Notably, we can trivially create wrappers for different scopes/features without threading optional arguments around. Also, they could potentially be combined (though I don't think we're likely to really need this functionality).

For example:

```py
class StaticFeatureExtractorFilter:
    def __init__(self, inner: StaticFeatureExtractor):
        self.inner = inner

    def __getattr__(self, name):
        if hasattr(self, name):
            # if the filter has an override, use that
            return getattr(self, name)
        # otherwise use the inner feature extractor
        return getattr(self.inner, name)

class FunctionFilter(StaticFeatureExtractorFilter):
    def __init__(self, inner: StaticFeatureExtractor, functions: Set[Address]):
        super().__init__(self, inner)
        self.functions = functions

    def get_functions(self):
        yield from (f for f in self.inner.get_functions() if f.address in self.functions)
```

Then we can use the filter like so:

```py
wanted_functions: Set[Address] = get_wanted_functions_from_cli(args)
if wanted_functions:
    # if the user wants to only show specific functions, we filter them down
    feature_extractor = FunctionFilter(feature_extractor, wanted_functions)
else:
    # otherwise, we use the full feature extractor
    pass

...

# use either the filtered extractor or the full extractor interchangably
find_capabilities(feature_extractor, ...)
```

Of course, we can filter processes/threads/basic blocks/etc. in the same way.

Would you be open to discuss alternative designs like this @yelhamer? Or any thoughts @mr-tz @mike-hunhoff

**Comment by yelhamer** [comment]
@williballenthin I really like the design you proposed and I'm willing to implement it. Would the filter classes reside in the base_extractor.py file?

Also, should we make the base "StaticFeatureExtractorFilter" a child of the "StaticFeatureExtractor" class? since I think there are some cases (if I am not mistaken) where we use the instance of the extractor we're passing to determine whether the analysis is static or dynamic.

**Comment by williballenthin** [comment]
> I think there are some cases (if I am not mistaken) where we use the instance of the extractor we're passing to determine whether the analysis is static or dynamic.

Good point. Also I think this will better satisfy mypy. 

OTOH, I'm not sure that the hasattr checks will work as written (since the base class has empty implementations of these) so that will take some tweaking. Should still be possible. 

We should also pass along the inner name appropriately, since I think the metadata structure includes the feature extractor name.

**Comment by yelhamer** [comment]
@williballenthin I have also thought of the following possible implementation using a function factory:

```python
def FunctionFilter(extractor: StaticFeatureExtractor, functions: Set) -> StaticFeatureExtractor:
    from types import MethodType

    get_functions = extractor.get_functions  # fetch original get_functions()

    def filtered_get_functions(self):
        yield from (f for f in get_functions() if f.address in functions)

    extractor.get_functions = MethodType(filtered_get_functions, extractor)
    return extractor
```

Then we can do as you suggested:

> ```python
> wanted_functions: Set[Address] = get_wanted_functions_from_cli(args)
> if wanted_functions:
>     # if the user wants to only show specific functions, we filter them down
>     feature_extractor = FunctionFilter(feature_extractor, wanted_functions)
> else:
>     # otherwise, we use the full feature extractor
>     pass
> 
> ...
> 
> # use either the filtered extractor or the full extractor interchangably
> find_capabilities(feature_extractor, ...)
> ```

Another question remains which is whether we want to register which filters an extractor has installed in it. If we want to do so then we might just consider storing the set of desired functions as an attribute in the extractor, then reference it internally in the get_functions() routine (without needing any wrapping or so).

**Comment by williballenthin** [comment]
> I have also thought of the following possible implementation using a function factory

This looks like it would also work. I'm not sure of the pros/cons right now, so perhaps try one of the new implementations and let's see how it feels?

File: tests/test_capabilities.py
Comment (williballenthin) [inline tests/test_capabilities.py:201]: Ah interesting. This is maybe a good reason not to mutate the extractor directly and wrap it instead.

Comment (yelhamer) [reply tests/test_capabilities.py:201]: Oh good point... Hmm I think I'll try to implement that instead.

Comment (williballenthin) [reply tests/test_capabilities.py:201]: thanks. that may have its own issues - lets see.

File: capa/main.py
Comment (mr-tz) [inline capa/main.py:285]: To discuss: is this needed besides show-features? if not, should we add it here?

**Comment by yelhamer** [comment]
@williballenthin @mr-tz , I have made some adjustments right now, please feel free to review them.

Also, this PR would make the Drakvuf extractor (and other extractors) faster, but I was also thinking of also applying the function/process filters to the extractors when they are constructed, so that non-essential functions/processes don't get loaded in the first place.

My main motive behind that is that the Drakvuf extractor consumes too much RAM upon initialization, and it would be nice to not create Pydantic models for the (winapi/native) calls that won't be used anyways. I am thinking of doing that (once both PRs I have open are merged) by passing the list of target processes to the feature extractor when it gets constructed (i.e., passing them to the get_extractor_from_cli() method down to the extractor.from_raw_report() method), and I was wondering whether the approach we have in this PR makes sense for that goal. In other words, do you think there could be a another approach to the problem of this PR that would allow us to filter unnecessary processes/functions upon extractor construction? or is it better that we keep the two separate?

Idk if what I said makes sense :sweat_smile:  , if it isn't then please let me know and I'll try to reword it.

**Comment by williballenthin** [comment]
I see what you're saying about extractors using the CLI target args so they don't have to do so much work. 

I think it's ok for the current strategy to work alongside extractors using the CLI args to do less work. Though I'd like to see how many changes are needed to thread this new information to the Drakvuf extractor. 

Alternatively, if only Drakvuf will usually use this targets functionality, I wonder if we can instead provide some utilities to pre-process the report so that it's optimized for capa.

File: capa/main.py
Comment (mr-tz) [reply capa/main.py:285]: @yelhamer what do you think here?

Comment (yelhamer) [reply capa/main.py:285]: @mr-tz I think using `--functions` for both main and show-features would be better, but I feel like that might be a breaking change however and I wouldn't like to delay the merge of this feature into the next release just because of that.

Maybe we could get this PR merged and then open a second PR (with a milestone for the next major release) that uses the `--functions` keyword instead of `-F`/`--function` in show-features, and then document the breaking change there?

Comment (mr-tz) [reply capa/main.py:285]: I like that. Let's remove it from this PR.

Comment (yelhamer) [reply capa/main.py:285]: do you mean remove `-F`/`--function` from show-features.py?

Comment (mr-tz) [reply capa/main.py:285]: I initially understood to remove the functions filter from this PR.

Comment (yelhamer) [reply capa/main.py:285]: Oh, the purpose of this PR is to have that process/function filtering in `show-features.py` applied to capability extractions as well.

I've implemented here (to capa's core) in a more general manner so it can be used by `show-features.py` without having to filter there.

Comment (mr-tz) [reply capa/main.py:285]: Where do you see the functions filter being used and helpful besides in show-features (my original comment here)?

Comment (mr-tz) [reply capa/main.py:285]: good points

**Comment by yelhamer** [comment]
Tests pass!

File: capa/main.py
Comment (williballenthin) [inline capa/main.py:292]:
```suggestion
            help="provide a list of comma-separated processes to analyze (dynamic analysis).",
```

File: capa/main.py
Comment (williballenthin) [inline capa/main.py:281]: I wonder if we should use a prefix for these arguments, to help indicate they all behave similarly. Perhaps `--restrict-to-functions` and `--restrict-to-processes`.

Furthermore, we *could* dynamically generate/parse these for all the scopes, rather than hard coding them here.

I'm also not sure whether or not we want to show these by default, or allow them to be hidden/expert options.

**Review by williballenthin — CHANGES_REQUESTED** [review]
I'm not yet happy with the way that the filter objects try to appear like the extractor objects. There's also some naming discussions that we should all resolve in this thread.

File: capa/features/extractors/base_extractor.py
Comment (yelhamer) [reply capa/features/extractors/base_extractor.py]: I have reverted to the previous implementation I had (modifying the extractor's functions). However, I tried to copy the extractor before I alter the method (essentially changing the method of the copy of the extractor object and returning that new copy). This gets around the issue we had before (tests not working properly since filters would alter the extractor object).

Another approach that might be good as well is to add a list of processes or functions to filter to the static or dynamic feature extractor abstract class, and then updating that list with each call to `ProcessFilter()` or `FunctionFilter()` respectively, but I am not sure if that's better than the current approach.

Let me know if you think that approach would be better (or if you have any other ideas!).

File: capa/main.py
Comment (yelhamer) [reply capa/main.py]: I've moved to this naming standard.

As for filtering other scopes, I am not sure whether that is desired or not now. Nevertheless, I have kept the room for adding any other filters in the future.

As for showing or hiding them, I lean more towards showing them since I feel like filtering by process/function is general and intuitive enough. Maybe make bb/insn or thread/call filtering hidden in the future (if we choose to add them).

File: capa/main.py
Comment (yelhamer) [reply capa/main.py]: I have updated this routine now. Please let me know if I need to do more here.

**Comment by mr-tz** [comment]
I'll defer to @williballenthin for final review and to comment on the maturity of this to be merged.

File: CHANGELOG.md
Comment (williballenthin) [inline CHANGELOG.md:6]:
```suggestion
- add the ability to select which specific functions or processes to analyze @yelhamer
```

File: CHANGELOG.md
Comment (williballenthin) [inline CHANGELOG.md:7]:
```suggestion
- cli: add the ability to select which specific functions or processes to analyze @yelhamer
```

File: capa/features/extractors/base_extractor.py
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:302]:
```suggestion
    original_get_functions = extractor.get_functions  # fetch original get_functions()
```

File: capa/features/extractors/base_extractor.py
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:308]: Rather than a comment that describes *what* is done, please add a comment that describes why you make this copy. I don't quite understand it yet, so this context would be useful to any future reader.

Comment (williballenthin) [reply capa/features/extractors/base_extractor.py:308]: and same comment below.

File: capa/main.py
Comment (williballenthin) [inline capa/main.py:295]:
```suggestion
            help="provide a list of comma-separated function virtual addresses to analyze (static analysis).",
```

File: capa/main.py
Comment (williballenthin) [inline capa/main.py:303]:
```suggestion
            help="provide a list of comma-separated process IDs to analyze (dynamic analysis).",
```

File: capa/main.py
Comment (williballenthin) [inline capa/main.py:810]: please create a type for the return value, or use `typing.TypedDict` to specify the expected keys in this dict.

Comment (williballenthin) [reply capa/main.py:810]:
```py
@dataclass
class FilterConfig:
    processes: set[int]
    functions: set[int]
    ...
```

File: capa/main.py
Comment (williballenthin) [inline capa/main.py:925]: i think this should be moved within `get_extractor_from_cli` so that all callers benefit from this filtering capability

Comment (williballenthin) [reply capa/main.py:925]: also, the `if any(...)` should go within `apply_extractor_filters` so that we can always call the routine unconditionally. this requires less cognitive overhead for readers/users.

File: capa/main.py
Comment (williballenthin) [inline capa/main.py:825]: should check to see that `functions` is present and non-empty. ditto with `processes`.

Comment (williballenthin) [reply capa/main.py:825]: because i'm thinking about the future where we may add additional filters, so that its not implicit that "if is static and there are filters then it must be a function filter".

File: capa/main.py
Comment (williballenthin) [inline capa/main.py:811]: we may want to factor out the `is_static_analysis` into its own routine. though its not required in this PR.

File: doc/usage.md
Comment (williballenthin) [inline doc/usage.md:14]:
```suggestion
Use the `--restrict-to-functions` option to extract capabilities from only a selected set of functions.
For example, `capa sample.exe --restrict-to-functions 0x4019C0,0x401CD0` will only extract the capabilities in the functions found at
```

File: doc/usage.md
Comment (williballenthin) [inline doc/usage.md:19]:
```suggestion
Use the `--restrict-to-processes` option to extract capabilities from only a selected set of processes.
For example, `capa report.log --restrict-to-processes 3888,3214,4299` will extract capabilities only from the processes 3888, 3214, and 4299.
```

Comment (williballenthin) [inline doc/usage.md:19]: how would a user know what the PIDs are? please suggest how a user might know about the process tree.

File: doc/usage.md
Comment (williballenthin) [inline doc/usage.md:13]:
```suggestion
Use the `--functions` option to extract capabilities from only a selected set of functions.
This is useful when XXX and we want YYY.
```

Comment (williballenthin) [reply doc/usage.md:13]: help users quickly understand when they should use this option (such as data overload).

File: capa/features/extractors/base_extractor.py
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:495]: overall this approach seems pretty reasonable to me. as long as tests work, im happy!

File: capa/main.py
Comment (williballenthin) [inline capa/main.py:290]: thanks for updating these names

**Review by williballenthin — CHANGES_REQUESTED** [review]
This implementation looks pretty good! I've left comments inline. Please address the comments tagged 👀 and optionally the remaining ones. We can merge this as soon as tomorrow assuming the points are addressed :-)

File: capa/features/extractors/base_extractor.py
Comment (yelhamer) [reply capa/features/extractors/base_extractor.py]: I've added a comment now. Please let me know if it's not clear enough.

File: capa/main.py
Comment (yelhamer) [reply capa/main.py:830]: > because i'm thinking about the future where we may add additional filters, so that its not implicit that "if is static and there are filters then it must be a function filter".

I've added this now. I am doing it with an assert now however, since I imagine that if we add a basic blocks filter for example then we would have an if-elif-else statement there or somewhere else to check the fields.

File: capa/main.py
Comment (yelhamer) [reply capa/main.py]: I have done this now. I am now getting the filters and applying them from within the `get_extractor_from_cli`.

File: doc/usage.md
Comment (yelhamer) [reply doc/usage.md]: I believe most sandboxes (at least the ones we support now) provide process trees as part of the output. I have said that now in the `usage.md` file.

**Comment by yelhamer** [comment]
@williballenthin I think I have addressed all of the the comments. Let me know if there's anything else I need to do :)

File: capa/main.py
Comment (williballenthin) [inline capa/main.py]:
```suggestion
    processes: Set[int]
    functions: Set[int]
```

indexing `set` as a type hint was added in Python 3.9, whereas we still support 3.8 for a little longer.
https://docs.python.org/3/library/typing.html#typing.Set

File: capa/features/extractors/base_extractor.py
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:310]: great!

File: capa/features/extractors/base_extractor.py
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:302]:
```suggestion
    original_get_functions = extractor.get_functions
```

File: capa/features/extractors/base_extractor.py
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:489]:
```suggestion
    original_get_processes = extractor.get_processes
```

File: capa/main.py
Comment (yelhamer) [reply capa/main.py]: I see. I have switched to `typing.Set` now.

File: capa/main.py
Comment (williballenthin) [inline capa/main.py:126]: great!

File: doc/usage.md
Comment (williballenthin) [inline doc/usage.md:21]:
```suggestion
generated from analyzing non-malicious processes that can be reported by some sandboxes, as well as reduce the execution time 
```

**Review by williballenthin — APPROVED** [review]
🥳

**Comment by williballenthin** [comment]
ah, CLA failure is cause GH associated the wrong email with my suggestions. well, its me, and i'm covered under the CLA.

**Comment by williballenthin** [comment]
let's go!

**Pull request merged**
